### PR TITLE
Add interface for classes generated for enum types and fix sealed classes json representation

### PR DESCRIPTION
**Pull request by lwasyl**
Objects representing enums in sealed classes don't have stable toString() implementation: instead, they'll return the class name along with memory address for that object. This breaks SQLite-based caching for queries that accept enum parameters, as toString() is used to generate the query key, so it would change between app restarts. This commit fixes this behavior by adding an EnumValue interface to types representing enums, and using its `rawValue` property to build proper cache key.

Fixes https://github.com/apollographql/apollo-android/issues/2775

### CODE REVIEW

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.apollographql.apollo.api
+
+/**
+ * Base class for an enum type when generating sealed Kotlin class
+ */
+abstract class SealedEnumType {
```
Comment (lwasyl) [inline apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt:6]: I kind of wanted to follow `ScalarType` naming, but of course can change it to whatever is more appropriate.

Also, @martinbonnin suggested using an interface (https://github.com/apollographql/apollo-android/issues/2775#issuecomment-734399072), but I wasn't sure it's okay to have a default method with `toString` implementation

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/EnumType.kt [hunk 2]
```diff
@@ -37,6 +38,12 @@ private val primaryConstructorSpec =
         .addParameter("rawValue", String::class)
         .build()
 
+private val sealedSubclassPrimaryConstructorSpec =
```
Comment (lwasyl) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/EnumType.kt:41]: The `primaryConstructorSpec` is used also when generating `enum class` and for `Unknown` case, so we need a new one that overrides the raw value

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt:6]: > I kind of wanted to follow ScalarType naming, but of course can change it to whatever is more appropriate.

Naming is hard... I'm personnaly not a fan of `ScalarType` as it acts more like a type-safe `Binding` or `Mapping` between a GraphQL type and a Kotlin one. So I don't mind not following the logic here. 

What about the below:
```
interface EnumValue {
  val rawValue: String
}
```

? This will be closest to the spec: https://spec.graphql.org/draft/#EnumValue

> And if the idea was to add a case here then of course I can also do that, it's just overriding toString feels like a good solution

It was the original intent. I feel it's dangerous to rely on `toString()` because it typically exposes to these kind of bugs that we're seeing here. The compiler cannot catch if a new type is added a someone forgets to override `toString()`. 

Ultimately, I was thinking of moving this:
```
fun writeToJson(value: Any?, jsonWriter: JsonWriter) {
```
to something more like:
```
fun writeToJson(value: InputValue, jsonWriter: JsonWriter) {
```

where `InputValue` would be an instance of an InputType as defined by https://spec.graphql.org/draft/#IsInputType(), i.e. scalar, enum or InputObject

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/EnumType.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/EnumType.kt:41]: Makes sense 👍

**Comment by martinbonnin** [comment]
Thanks a lot for looking into this! I took the liberty of pushing a small integration test to your branch, I hope you don't mind. The test checks that the cache key is the one we expect. I didn't think of any other test to do since we can't really restart the JVM in the middle of the tests to make the cache key change...

File: apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt (hunk 1 above)
Comment (lwasyl) [reply apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/SealedEnumType.kt]: I changed the abstract class to `interface EnumValue`, following the GraphQL spec definitely makes sense! Also with that change I now could implement that interface for all enum classes, not only the ones generated as sealed classes.

`writeToJson(value: InputValue` also makes perfect sense. I originally thought it's more of a general-purpose utility, that's why I had doubts :)

**Comment by lwasyl** [comment]
> I took the liberty of pushing a small integration test to your branch, I hope you don't mind. The test checks that the cache key is the one we expect. 

Not at all, now at least I know for sure the fix works ;)

> I didn't think of any other test to do since we can't really restart the JVM in the middle of the tests to make the cache key change...

Technically another `object` instance could be created using reflection, or by serializing and deserializing an object instance. I don't think that's necessary though, I don't see a scenario where such test would fail but yours would pass

File: apollo-compiler/src/test/graphql/com/example/java_beans_semantic_naming/type/Episode.kt [hunk 3]
```diff
@@ -5,15 +5,16 @@
...
 import kotlin.String
 
 /**
  * The episodes in the Star Wars trilogy
  */
 enum class Episode(
-  val rawValue: String
-) {
+  override val rawValue: String
+) : EnumValue {
```
Comment (martinbonnin) [inline apollo-compiler/src/test/graphql/com/example/java_beans_semantic_naming/type/Episode.kt:17]: Paranoid checks: I think this is a backward compatible change but since it touches regular enums and not only sealed classes this has the potentially to impact a lot more users. Are we missing possibly missing a weird edge case?

Comment (lwasyl) [reply apollo-compiler/src/test/graphql/com/example/java_beans_semantic_naming/type/Episode.kt:17]: I don't see any 🤔 One edge case would be using jar with generated old classes (no interface) with new runtime, but it should work (just the check won't trigger). For enum classes behavior doesn't change at all, as from what I see both raw value and enum name always match (so cache won't break when updating app with new generated classes). 

I can of course revert this if you find some problem or just feel I should only fix the bug with sealed classes and let enum classes as is. It just felt weird leaving actual enum classes without having them implement `EnumValue` interface 🙂

Comment (martinbonnin) [reply apollo-compiler/src/test/graphql/com/example/java_beans_semantic_naming/type/Episode.kt:17]: Yep, I think it's safe too 👍 and way more uniform like this 👍 .

**Review by martinbonnin — APPROVED** [review]

**Comment by lwasyl** [comment]
@martinbonnin the CI seems to be acting up so I took this opportunity and fixed commit message (unless CI isn't supposed to pass for PRs from forks). So just fyi I didn't push any more changes, only changed the commit message

**Comment by martinbonnin** [comment]
Yea, the CI randomly hangs for 6 hours 😒 . Sorry about that. It looks like it's happening in integration test but we haven't been able to find what's triggering the hang. Of course it's highly non reproducible...

I'm starting to log failures in this PR in the hopes of finding a common pattern: https://github.com/apollographql/apollo-android/pull/2764.
